kategory/src/main/kotlin/kategory/instances/WriterTInstances.kt

```diff
@@ -1,31 +1,104 @@
+            object : WriterTFunctor<F, W> {
+                override fun F0(): Functor<F> = FF
+            }
+
+    fun <F, W> monoidKForWriterT(MF: Monad<F>, MKF: MonoidK<F>): MonoidK<WriterTKindPartial<F, W>> =
+            object : WriterTMonoidK<F, W> {
+                override fun MF(): Monad<F> = MF
+                override fun F0(): MonoidK<F> = MKF
+            }
+}
+
+interface WriterTInstances3 {
+
+    fun <F, W> semigroupKForWriterT(MF: Monad<F>, MKF: SemigroupK<F>): SemigroupK<WriterTKindPartial<F, W>> =
+            object : WriterTSemigroupK<F, W> {
+                override fun MF(): Monad<F> = MF
+                override fun F0(): SemigroupK<F> = MKF
+            }
+}
+
+interface WriterTApplicative<F, W> : Applicative<WriterTKindPartial<F, W>>, WriterTFunctor<F, W> {
+
+    override fun F0(): Monad<F>
+    fun L0(): Monoid<W>
+
+    override fun <A> pure(a: A): HK<WriterTKindPartial<F, W>, A> = WriterT(F0(), F0().pure(L0().empty() toT a))
 
-    override fun <A, B> flatMap(fa: WriterTKind<F, W, A>, f: (A) -> HK<WriterTKindPartial<F, W>, B>): WriterT<F, W, B> = fa.ev().flatMap({ f(it).ev() }, SG())
+    override fun <A, B> ap(fa: HK<WriterTKindPartial<F, W>, A>, ff: HK<WriterTKindPartial<F, W>, (A) -> B>): HK<WriterTKindPartial<F, W>, B> =
+            ap(fa, ff)
+
+    override fun <A, B> map(fa: HK<WriterTKindPartial<F, W>, A>, f: (A) -> B): HK<WriterTKindPartial<F, W>, B> {
+        return super<WriterTFunctor>.map(fa, f)
+    }
+}
+
+interface WriterTMonad<F, W> : WriterTApplicative<F, W>, Monad<WriterTKindPartial<F, W>> {
+    override fun <A, B> flatMap(fa: WriterTKind<F, W, A>, f: (A) -> HK<WriterTKindPartial<F, W>, B>): WriterT<F, W, B> = fa.ev().flatMap({ f(it).ev() }, L0())
 
     override fun <A, B> tailRecM(a: A, f: (A) -> HK<WriterTKindPartial<F, W>, Either<A, B>>): WriterT<F, W, B> =
-            WriterT(MM(), MM().tailRecM(a, {
-                MM().map(f(it).ev().value) {
+            WriterT(F0(), F0().tailRecM(a, {
+                F0().map(f(it).ev().value) {
                     when (it.b) {
                         is Either.Left<A, B> -> Either.Left(it.b.a)
                         is Either.Right<A, B> -> Either.Right(it.a toT it.b.b)
                     }
                 }
             }))
 
+    override fun <A, B> ap(fa: HK<WriterTKindPartial<F, W>, A>, ff: HK<WriterTKindPartial<F, W>, (A) -> B>): HK<WriterTKindPartial<F, W>, B> {
+        return super<Monad>.ap(fa, ff)
+    }
+}
+
+interface WriterTFunctor<F, W> : Functor<WriterTKindPartial<F, W>> {
+    fun F0(): Functor<F>
+
+    override fun <A, B> map(fa: HK<WriterTKindPartial<F, W>, A>, f: (A) -> B): HK<WriterTKindPartial<F, W>, B> = fa.ev().map { f(it) }
+}
+
+interface WriterTMonadFilter<F, W> : WriterTMonad<F, W>, MonadFilter<WriterTKindPartial<F, W>> {
+    override fun F0(): MonadFilter<F>
+
+    override fun <A> empty(): HK<WriterTKindPartial<F, W>, A> = WriterT(F0(), F0().empty())
```

Comment:
This method an all others that may return HK repesentations can use just `WriterT<F, W, A>` as return type.